### ISSUE TITLE
A result problem when running [10obs, 8cores] environment

### ISSUE THREAD

**Issue by FANnn100** [RobustFieldAutonomyLab/Distributional_RL_Navigation#3]
Hello, Thank you for your work. I encountered an issue while running the run experiments file. Without modifying any settings, the experimental results for the complex environment [10obs, 8cores] are as follows:

=== Finish 500 experiments ===
adaptive_IQN | success rate: 0.71 | out of area rate: 0.12 | avg_time: 37.75 | avg_energy: 97.43 | avg_compute_t: 0.0010611089374018853
IQN_0.25 | success rate: 0.49 | out of area rate: 0.41 | avg_time: 38.70 | avg_energy: 99.71 | avg_compute_t: 0.001004070807167153
IQN_0.5 | success rate: 0.66 | out of area rate: 0.24 | avg_time: 37.52 | avg_energy: 94.62 | avg_compute_t: 0.001005096895758446
IQN_0.75 | success rate: 0.70 | out of area rate: 0.15 | avg_time: 38.17 | avg_energy: 97.55 | avg_compute_t: 0.001001747665702137
IQN_1.0 | success rate: 0.75 | out of area rate: 0.07 | avg_time: 37.78 | avg_energy: 97.41 | avg_compute_t: 0.0009959756102798688
DQN | success rate: 0.61 | out of area rate: 0.00 | avg_time: 33.05 | avg_energy: 96.16 | avg_compute_t: 0.0001960611697107753
APF | success rate: 0.58 | out of area rate: 0.07 | avg_time: 53.13 | avg_energy: 162.81 | avg_compute_t: 4.319793696453127e-05
BA | success rate: 0.37 | out of area rate: 0.01 | avg_time: 49.67 | avg_energy: 132.89 | avg_compute_t: 3.5731663451286436e-05

This doesn't seem to meet our expectations. Could you please let me know if there is any problem? :)

**Comment by bexilin** [comment]
Hi, random quantile samples are drawn from the uniform distribution before each forward pass of IQN, and thus introduce randomness in IQN's behavior and statistics in experiments. You may try using a fixed set of quantile samples to get a more deterministic behavior. As an example, you can modify line # 138-153 in thirdparty/IQN/model.py in the following way.

```
       # temp for reproducibility
       self.taus = None

 def calc_cos(self, batch_size, n_tau=8, cvar=1.0):
        """
        Calculating the cosinus values depending on the number of tau samples
        """
        # temp for reproducibility
        if self.taus is None:
            t = np.arange(0.05,1,0.03).astype(np.float32)
            self.taus = torch.from_numpy(t).to(self.device).unsqueeze(-1) # (batch_size, n_tau, 1) for broadcast
        # taus = torch.rand(batch_size, n_tau).to(self.device).unsqueeze(-1) # (batch_size, n_tau, 1) for broadcast
        #print(taus)

        # distorted quantile sampling
        taus = self.taus * cvar
```

**Comment by FANnn100** [comment]
Thanks for this reply. I follow this instruction and rerun the run_experiments.py, the result can be

`=== Finish 500 experiments ===
adaptive_IQN | success rate: 0.77 | out of area rate: 0.08 | avg_time: 37.75 | avg_energy: 97.72 | avg_compute_t: 0.0010383376706107875
IQN_0.25 | success rate: 0.52 | out of area rate: 0.35 | avg_time: 38.84 | avg_energy: 99.32 | avg_compute_t: 0.000999825670639914
IQN_0.5 | success rate: 0.69 | out of area rate: 0.19 | avg_time: 37.94 | avg_energy: 95.34 | avg_compute_t: 0.001001367683986533
IQN_0.75 | success rate: 0.76 | out of area rate: 0.10 | avg_time: 37.47 | avg_energy: 94.35 | avg_compute_t: 0.0009948614683529572
IQN_1.0 | success rate: 0.78 | out of area rate: 0.03 | avg_time: 36.35 | avg_energy: 92.96 | avg_compute_t: 0.0009973955625858463
DQN | success rate: 0.61 | out of area rate: 0.00 | avg_time: 33.05 | avg_energy: 96.16 | avg_compute_t: 0.00020211031537429536
APF | success rate: 0.58 | out of area rate: 0.07 | avg_time: 53.13 | avg_energy: 162.81 | avg_compute_t: 4.092847572925941e-05
BA | success rate: 0.37 | out of area rate: 0.01 | avg_time: 49.67 | avg_energy: 132.89 | avg_compute_t: 3.648707112149082e-05`

It's still hard for adaptive_IQN to be better than IQN_1.0, it's quite weird. If I use the same fixed seed and the same taus as used in the paper, I should reach the same experimental results, right? :open_mouth:

**Comment by bexilin** [comment]
Hi, the data we gathered during experiments for IQN used random quantile samples instead of the deterministic ones. Besides, from my knowledge the results can vary in different machines due to differences in numerical precision.

**Comment by FANnn100** [comment]
Oh, I see. Maybe I should train the IQN by myself on my device instead of using the pre-trained model, right?:dizzy_face:

**Comment by bexilin** [comment]
I double checked the pre-trained models and found out that they were actually in an older version. Now the pre-trained models have been replaced with the correction version, and the same experiment results as those in the paper can be reproduced on my device (using the fixed set of quantile samples mentioned above). Thanks for your feedback in this issue.

**Comment by FANnn100** [comment]
Yes, I rerun the experiments, It shows consistent result with the part of the paper. I think this work is truly helpful. Thanks :blush:
